backend/src/resolvers/user_mutations.js

```diff
@@ -0,0 +1,40 @@
+const hashing = require("../services/hashing.js");
+const jwt = require("../services/jwt.js");
+import { AuthenticationError, UserInputError } from "apollo-server";
+
+function isPasswordStrong(password) {
+  return password.length >= 8;
+}
+
+export const usersMutationResolver = {
+  Mutation: {
+    login: async (parent, args, context) => {
+      let user = await context.dataSources.db.getUserByEmail(args.email);
+      if (!user)
+        throw new UserInputError(
+          "There is no user registered with this Email!"
+        );
+      let isPasswordCorrect = await hashing.compare(
+        args.password,
+        user.password
+      );
+      if (!isPasswordCorrect)
+        throw new AuthenticationError("Password did not match!");
+      return jwt.issueToken(user.id);
+    },
+
+    signup: async (parent, args, context) => {
+      let emailExists = await context.dataSources.db.emailExists(args.email);
+      if (emailExists) throw new UserInputError("Email already exists!");
+      if (!isPasswordStrong(args.password))
+        throw new UserInputError("Password must be atleast 8 characters!");
+      let encryptedPassword = await hashing.hash(args.password);
+      let user = await context.dataSources.db.addNewUser({
+        name: args.name,
+        email: args.email,
+        password: encryptedPassword,
+      });
+      return jwt.issueToken(user.id);
+    },
+  },
```

Comment:
denke, diese beiden Funktionen sollen in db.js definiert werden. Es wird sauberer sein. Oder?

Reply:
Ich finde es besser hier als der DB, aber ich bin mir auch nicht 100% sicher, wie die Trennung sein sollte. Die DB sollte nur Daten speichern, hashing und alles andere sollte dieser egal sein. Zuständig dafür wäre ein anderer Service - `login_service.js` oder etwas ähnliches